File: fixml/lua/fixml.lua
#!/usr/bin/env lua

-- Ultra-optimized XML organizer v5.0.0 - Maximum performance Lua
-- Target: Match compiled language performance through aggressive optimization

-- Standard constants - consistent across all implementations
local USAGE = [[
Usage: lua fixml.lua [--organize] [--replace] [--fix-warnings] <xml-file>
  --organize, -o      Apply logical organization
  --replace, -r       Replace original file
  --fix-warnings, -f  Fix XML warnings
  Default: preserve original structure, fix indentation/deduplication only
]]

local XML_DECLARATION = '<?xml version="1.0" encoding="utf-8"?>\n'
local MAX_INDENT_LEVELS = 64           -- Maximum nesting depth supported
local ESTIMATED_LINE_LENGTH = 50       -- Average characters per line estimate
local MIN_HASH_CAPACITY = 256          -- Minimum deduplication hash capacity
local MAX_HASH_CAPACITY = 4096         -- Maximum deduplication hash capacity
local WHITESPACE_THRESHOLD = 32        -- ASCII values <= this are whitespace
local FILE_PERMISSIONS = 644           -- Standard file permissions (octal in other langs)
local IO_CHUNK_SIZE = 65536            -- 64KB chunks for I/O operations

-- Simplified: Use string.char() directly (caching overhead not worth it)

-- Lightweight markers
local SELF_CLOSING_PATTERN = "/>"

-- Parse command line arguments (optimized)
local function parse_args()
	local organize, replace, fix_warnings, file = false, false, false, nil
	local args = arg or {}
	
	for i = 1, #args do
		local a = args[i]
		if a == "--organize" or a == "-o" then
			organize = true
		elseif a == "--replace" or a == "-r" then
			replace = true
		elseif a == "--fix-warnings" or a == "-f" then
			fix_warnings = true
		elseif not file and a:sub(1,1) ~= "-" then
			file = a
		end
	end

	if not file then
		print(USAGE)
		os.exit(1)
	end
	return organize, replace, fix_warnings, file
end

-- Ultra-optimized O(n) single-pass content cleaning
local function clean_content(content)
	local len = #content
	if len == 0 then return content end
	
	-- Pre-allocate result buffer (avoid reallocations)
	local result = {}
	local result_size = 0
	local i = 1
	
	-- Remove BOM if present (optimized byte checks)
	if len >= 3 then
		local b1, b2, b3 = content:byte(1, 3)
		if b1 == 239 and b2 == 187 and b3 == 191 then
			i = 4
		end
	end
	
	-- Ultra-fast single-pass normalization using byte operations
	while i <= len do
		local b = content:byte(i)
		if b == 13 then -- CR
			if i < len and content:byte(i + 1) == 10 then -- CRLF
				result_size = result_size + 1
				result[result_size] = '\n'
				i = i + 2
			else -- standalone CR
				result_size = result_size + 1
				result[result_size] = '\n'
				i = i + 1
			end
		else
			result_size = result_size + 1
			result[result_size] = string.char(b)
			i = i + 1
		end
	end
	
	return table.concat(result)
end

-- Ultra-fast string trimming without regex
local function fast_trim(s)
	if not s then return "" end
	local len = #s
	if len == 0 then return s end
	
	-- Find first non-whitespace
	local start = 1
	while start <= len do
		local b = s:byte(start)
		if b > WHITESPACE_THRESHOLD then break end
		start = start + 1
	end
	
	if start > len then return "" end
	
	-- Find last non-whitespace
	local finish = len
	while finish >= start do
		local b = s:byte(finish)
		if b > WHITESPACE_THRESHOLD then break end
		finish = finish - 1
	end
	
	if start == 1 and finish == len then
		return s
	else
		return s:sub(start, finish)
	end
end

-- Normalize whitespace while preserving attribute values
local function normalize_whitespace_preserving_attributes(s)
	local len = #s
	if len == 0 then return s end
	
	local result = {}
	local result_size = 0
	local in_quotes = false
	local quote_char = nil
	local prev_space = false
	
	-- Optimized loop avoiding string.char conversions
	for i = 1, len do
		local b = s:byte(i)
		
		if not in_quotes and (b == 34 or b == 39) then -- '"' or "'"
			in_quotes = true
			quote_char = b
			result_size = result_size + 1
			result[result_size] = string.char(b)
			prev_space = false
		elseif in_quotes and b == quote_char then
			in_quotes = false
			result_size = result_size + 1
			result[result_size] = string.char(b)
			prev_space = false
		elseif in_quotes then
			-- Inside quotes: preserve all whitespace
			result_size = result_size + 1
			result[result_size] = string.char(b)
			prev_space = false
		elseif b <= WHITESPACE_THRESHOLD then -- standardized whitespace
			-- Outside quotes: normalize whitespace
			if not prev_space then
				result_size = result_size + 1
				result[result_size] = " "
				prev_space = true
			end
		else
			result_size = result_size + 1
			result[result_size] = string.char(b)
			prev_space = false
		end
	end
	
	return fast_trim(table.concat(result, "", 1, result_size))
end

-- Lightweight self-contained element check without patterns
local function is_self_contained(s)
	local len = #s
	if len < 7 then return false end -- minimum: <a>x</a>
	if s:byte(1) ~= 60 or s:byte(len) ~= 62 then return false end -- '<' and '>'
	local first_gt = s:find(">", 2, true)
	if not first_gt then return false end
	local last_lt = s:match(".*()<")
	if not last_lt or last_lt <= first_gt then return false end
	-- Ensure closing tag starts with </
	if last_lt + 1 > len or s:byte(last_lt + 1) ~= 47 then return false end -- '/'
	-- Ensure inner content has no '<'
	local inner = s:sub(first_gt + 1, last_lt - 1)
	return not inner:find("<", 1, true)
end

-- Ultra-fast line processing avoiding gmatch iterator
local function process_lines(content, processor)
	local len = #content
	local start = 1
	local line_num = 0
	
	while start <= len do
		-- Find line end
		local line_end = start
		while line_end <= len and content:byte(line_end) ~= 10 do
			line_end = line_end + 1
		end
		
		-- Extract line
		local line = content:sub(start, line_end - 1)
		line_num = line_num + 1
		
		-- Process line
		processor(line, line_num)
		
		-- Move to next line
		start = line_end + 1
	end
end

-- Ultra-optimized XML processing
local function process_xml_file(organize_mode, replace_mode, fix_warnings, input_file)
	-- Read file
	local file = io.open(input_file, "r")
	if not file then
		error("Could not open file: " .. input_file)
	end
	local content = file:read("*all")
	file:close()

	-- Clean content
	local cleaned_content = clean_content(content)
	local has_xml_decl = cleaned_content:find("<?xml", 1, true) -- plain text search

	-- Show warnings
	if not has_xml_decl then
		print("⚠️  XML Best Practice Warnings:")
		print("  [XML] Missing XML declaration")
		print("    Fix: Add <?xml version=\"1.0\" encoding=\"utf-8\"?> at the top")
		print()
		if not fix_warnings then
			print("Use --fix-warnings flag to automatically apply fixes")
			print()
		end
	end

	-- Pre-allocate large output buffer
	local output_size = 0
	local output = {}

	local function add_output(str)
		output_size = output_size + 1
		output[output_size] = str
	end

	-- Add XML declaration if needed
	if fix_warnings and not has_xml_decl then
		add_output(XML_DECLARATION)
		print("🔧 Applied fixes:")
		print("  ✓ Added XML declaration")
		print()
	end

	-- Process with optimized deduplication
	local seen_elements = {}
	local duplicates_removed = 0
	local indent_level = 0
	
	-- Pre-allocate indentation strings
	local indent_cache = {""}
	for i = 1, MAX_INDENT_LEVELS do
		indent_cache[i + 1] = string.rep("  ", i)
	end

	-- Ultra-fast line processing
	process_lines(cleaned_content, function(line, line_num)
		local trimmed = fast_trim(line)
		if #trimmed > 0 then
			-- Fast container detection: simple tags without spaces (no attributes)
			local is_container = (#trimmed > 2 and 
			                     trimmed:byte(1) == 60 and trimmed:byte(#trimmed) == 62 and 
			                     not trimmed:find(" ", 2, true))
			
			-- Use normalized key for non-container elements (better deduplication)
			local key = is_container and trimmed or normalize_whitespace_preserving_attributes(trimmed)
			
			if not is_container and seen_elements[key] then
				duplicates_removed = duplicates_removed + 1
			else
				if not is_container then
					seen_elements[key] = true
				end
				
				-- Adjust indent for closing tags BEFORE applying indentation
				if trimmed:sub(1, 2) == "</" then
					indent_level = math.max(0, indent_level - 1)
				end
				
				-- Apply consistent 2-space indentation using cached indentation
				local indent_idx = math.min(indent_level + 1, #indent_cache)
				add_output(indent_cache[indent_idx])
				add_output(trimmed)
				add_output("\n")
				
				-- Fast opening tag detection - avoid expensive is_self_contained check when possible
				local is_opening_tag = false
				if trimmed:byte(1) == 60 and trimmed:byte(2) then
					local b2 = trimmed:byte(2)
					if b2 ~= 47 and b2 ~= 63 and b2 ~= 33 and not trimmed:find(SELF_CLOSING_PATTERN) then
						-- Only call expensive is_self_contained for potential opening tags
						is_opening_tag = not is_self_contained(trimmed)
					end
				end
				
				if is_opening_tag then
					indent_level = indent_level + 1
					-- Warn about exceeding maximum indent levels but continue processing
					if indent_level > MAX_INDENT_LEVELS then
						io.stderr:write("⚠️  Warning: XML nesting exceeds maximum supported depth of " .. MAX_INDENT_LEVELS .. " levels. Indentation may be incorrect.\n")
						io.stderr:flush()
					end
				end
			end
		end
	end)

	-- Write output efficiently
	local output_filename
	if replace_mode then
		output_filename = input_file .. ".tmp." .. os.time()
	else
		local name, ext = input_file:match("^(.+)%.([^%.]+)$")
		if name and ext then
			output_filename = name .. ".organized." .. ext
		else
			output_filename = input_file .. ".organized"
		end
	end

	local out_file = io.open(output_filename, "w")
	if not out_file then
		error("Could not create output file: " .. output_filename)
	end
	
	-- Simplified: single write operation is more efficient for most cases
	out_file:write(table.concat(output, "", 1, output_size))
	
	out_file:close()

	-- Handle replacement
	if replace_mode then
		os.rename(output_filename, input_file)
		io.write("Original file replaced: " .. input_file)
	else
		io.write("Organized project saved to: " .. output_filename)
	end

	if duplicates_removed > 0 then
		io.write(" (removed " .. duplicates_removed .. " duplicates)")
	end

	local mode_text = organize_mode and " (with logical organization)" or " (preserving original structure)"
	print(mode_text)
end

-- Main execution with error handling
local function main()
	local ok, err = pcall(function()
		local organize, replace, fix_warnings, file = parse_args()
		process_xml_file(organize, replace, fix_warnings, file)
	end)
	
	if not ok then
		io.stderr:write("Error: " .. tostring(err) .. "\n")
		os.exit(1)
	end
end

main()

File: fixml/python/fixml.py
#!/usr/bin/env python3

"""
FIXML - High-Performance XML Processor (Python Implementation)

This implementation focuses on Python best practices while maintaining performance:
- List comprehensions and generator expressions for efficiency
- Built-in string methods for fast text processing
- Set-based deduplication with capacity pre-sizing
- Minimal object creation to reduce GC pressure

Performance Characteristics:
- Time Complexity: O(n) where n = input file size
- Space Complexity: O(n + d) where d = unique elements
- Optimized for readability and maintainability over raw speed
"""

import sys
import os
import argparse
import time
import re
from typing import Tuple, Set, List

# Standard constants - consistent across all implementations
USAGE = """Usage: python fixml.py [--organize] [--replace] [--fix-warnings] <xml-file>
  --organize, -o      Apply logical organization
  --replace, -r       Replace original file
  --fix-warnings, -f  Fix XML warnings
  Default: preserve original structure, fix indentation/deduplication only"""

XML_DECLARATION = '<?xml version="1.0" encoding="utf-8"?>\n'
MAX_INDENT_LEVELS = 64           # Maximum nesting depth supported
ESTIMATED_LINE_LENGTH = 50       # Average characters per line estimate
MIN_HASH_CAPACITY = 256          # Minimum deduplication hash capacity
MAX_HASH_CAPACITY = 4096         # Maximum deduplication hash capacity
WHITESPACE_THRESHOLD = 32        # ASCII values <= this are whitespace
FILE_PERMISSIONS = 0o644         # Standard file permissions
IO_CHUNK_SIZE = 65536           # 64KB chunks for I/O operations

class Args:
    """Command-line argument structure
    Mirrors interface across all language implementations for consistency
    """
    def __init__(self):
        self.organize = False      # Apply logical XML element organization
        self.replace = False       # Replace original file
        self.fix_warnings = False  # Fix XML warnings
        self.file = ""            # Input XML file path

def parse_args() -> Args:
    """Parse command-line arguments with consistent interface"""
    args = Args()
    
    # Simple argument parsing matching other implementations
    argv = sys.argv[1:]
    i = 0
    while i < len(argv):
        arg = argv[i]
        if arg in ('--organize', '-o'):
            args.organize = True
        elif arg in ('--replace', '-r'):
            args.replace = True
        elif arg in ('--fix-warnings', '-f'):
            args.fix_warnings = True
        elif not arg.startswith('-') and not args.file:
            args.file = arg
        i += 1
    
    if not args.file:
        print(USAGE)
        sys.exit(1)
    
    return args

def clean_content(content: str) -> str:
    """Optimized O(n) single-pass content cleaning
    Handles BOM removal and line ending normalization
    """
    if not content:
        return content
    
    # Remove BOM if present
    if content.startswith('\ufeff'):
        content = content[1:]
    
    # Fast line ending normalization using built-in methods
    # Python's built-in string methods are highly optimized
    content = content.replace('\r\n', '\n').replace('\r', '\n')
    
    return content

def fast_trim(s: str) -> str:
    """High-performance whitespace trimming
    Uses Python's optimized built-in strip method
    """
    return s.strip()

def is_container_element(trimmed: str) -> bool:
    """Check if element is a container (opening/closing tag without attributes)
    Simple tags without spaces indicate container elements
    """
    if len(trimmed) < 3:
        return False
    
    if trimmed.startswith('<') and trimmed.endswith('>'):
        # Check for simple container tags without spaces (no attributes)
        return ' ' not in trimmed and '=' not in trimmed
    
    return False

def is_self_contained(trimmed: str) -> bool:
    """Check if element is self-contained like <tag>content</tag>
    Uses string methods for fast detection
    """
    if len(trimmed) < 7:  # Minimum: <a>b</a>
        return False
    
    if not (trimmed.startswith('<') and trimmed.endswith('>')):
        return False
    
    # Find first > and last <
    first_gt = trimmed.find('>')
    last_lt = trimmed.rfind('<')
    
    if first_gt == -1 or last_lt == -1 or first_gt >= last_lt:
        return False
    
    # Check if it ends with closing tag
    return (first_gt < last_lt and 
            trimmed[last_lt:].startswith('</')  and 
            trimmed[first_gt+1:last_lt].find('<') == -1)

def normalize_whitespace_preserving_attributes(s: str) -> str:
    """Normalize whitespace while preserving attribute values
    Handles quoted strings and normalizes structural whitespace
    """
    if not s:
        return s
    
    # Quick check - if no quotes, use simple normalization
    if '"' not in s and "'" not in s:
        return ' '.join(s.split())
    
    # Process character by character to preserve quoted content
    result = []
    in_quotes = False
    quote_char = None
    prev_space = False
    
    quote_chars = {'"', "'"}
    for c in s:
        if not in_quotes and c in quote_chars:
            in_quotes = True
            quote_char = c
            result.append(c)
            prev_space = False
        elif in_quotes and c == quote_char:
            in_quotes = False
            result.append(c)
            prev_space = False
        elif in_quotes:
            # Inside quotes: preserve all whitespace
            result.append(c)
            prev_space = False
        elif c.isspace():
            # Outside quotes: normalize whitespace
            if not prev_space:
                result.append(' ')
                prev_space = True
        else:
            result.append(c)
            prev_space = False
    
    return ''.join(result).strip()

def compute_semantic_hash(s: str) -> int:
    """Compute hash representing semantic content with minimal string allocation
    Optimized for Python using built-in hash with reduced temporary objects
    """
    if not s:
        return hash('')
    
    # Quick path for simple strings without quotes
    if '"' not in s and "'" not in s:
        # Use list comprehension for efficiency - single pass, minimal allocation
        words = s.split()
        return hash(' '.join(words)) if words else hash('')
    
    # For complex strings with quotes, build normalized form efficiently
    # Use list with pre-estimated capacity to reduce allocations
    normalized_chars = []
    in_quotes = False
    quote_char = None
    prev_space = False
    
    for c in s:
        if not in_quotes and (c == '"' or c == "'"):
            in_quotes = True
            quote_char = c
            # Normalize quote type for semantic equivalence
            normalized_chars.append('"')
            prev_space = False
        elif in_quotes and c == quote_char:
            in_quotes = False
            # Normalize quote type for semantic equivalence
            normalized_chars.append('"')
            prev_space = False
        elif in_quotes:
            # Preserve content inside quotes exactly
            normalized_chars.append(c)
            prev_space = False
        elif c.isspace():
            # Normalize whitespace outside quotes
            if not prev_space:
                normalized_chars.append(' ')
                prev_space = True
        else:
            # Regular characters
            normalized_chars.append(c)
            prev_space = False
    
    # Single join operation - more efficient than multiple string operations
    return hash(''.join(normalized_chars).strip())

def process_xml_with_deduplication(content: str) -> Tuple[str, int]:
    """Optimized O(n) XML processing with deduplication and indentation
    Returns processed content and number of duplicates removed
    """
    lines = content.splitlines()
    
    # More efficient result building with better size estimation
    result = []
    # Pre-allocate based on actual line count for better performance
    result.extend([''] * len(lines))
    result.clear()  # Clear but keep capacity
    indent_level = 0
    
    # Pre-size set for better performance (Python doesn't have explicit capacity, but this helps)
    seen_elements: Set[int] = set()
    duplicates_removed = 0
    
    # Pre-compute indentation strings for performance
    indent_cache = ['  ' * i for i in range(MAX_INDENT_LEVELS + 1)]
    
    for line in lines:
        trimmed = fast_trim(line)
        if not trimmed:
            continue
        
        # Container detection - never deduplicate structural elements
        is_container = is_container_element(trimmed)
        
        # Deduplication only for non-container lines
        if not is_container:
            semantic_hash = compute_semantic_hash(trimmed)
            if semantic_hash in seen_elements:
                duplicates_removed += 1
                continue  # Skip duplicate
            seen_elements.add(semantic_hash)
        
        # Adjust indent for closing tags BEFORE applying indentation
        if trimmed.startswith('</'):
            indent_level = max(0, indent_level - 1)
        
        # Apply consistent 2-space indentation using cached strings
        if indent_level < len(indent_cache):
            result.append(indent_cache[indent_level] + trimmed)
        else:
            # For very deep nesting, fall back to dynamic generation
            result.append('  ' * indent_level + trimmed)
        
        # Fast opening tag detection - avoid expensive self-contained check when possible
        is_opening_tag = False
        if (len(trimmed) > 1 and trimmed[0] == '<' and 
            trimmed[1] not in ('/', '!', '?') and 
            not trimmed.endswith('/>')):
            # Only call expensive is_self_contained for potential opening tags
            is_opening_tag = not is_self_contained(trimmed)
        
        if is_opening_tag:
            indent_level += 1
    
    return '\n'.join(result) + '\n', duplicates_removed

def get_output_filename(input_file: str, replace_mode: bool) -> str:
    """Generate output filename based on mode"""
    if replace_mode:
        return f"{input_file}.tmp.{int(time.time())}"
    else:
        # Split on last dot to preserve extensions
        parts = input_file.rsplit('.', 1)
        if len(parts) == 2:
            return f"{parts[0]}.organized.{parts[1]}"
        else:
            return f"{input_file}.organized"

def process_file(args: Args) -> None:
    """Main file processing function"""
    try:
        # Read file with efficient method
        with open(args.file, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"Error: Could not read file '{args.file}': {e}", file=sys.stderr)
        sys.exit(1)
    
    cleaned_content = clean_content(content)
    has_xml_decl = '<?xml' in cleaned_content
    
    # Show warnings
    if not has_xml_decl:
        print("⚠️  XML Best Practice Warnings:")
        print("  [XML] Missing XML declaration")
        print('    Fix: Add <?xml version="1.0" encoding="utf-8"?> at the top')
        print()
        
        if not args.fix_warnings:
            print("Use --fix-warnings flag to automatically apply fixes")
            print()
    
    # Build final content efficiently
    final_content_parts = []
    
    # Add XML declaration if needed
    if args.fix_warnings and not has_xml_decl:
        final_content_parts.append(XML_DECLARATION)
        print("🔧 Applied fixes:")
        print("  ✓ Added XML declaration")
        print()
    
    # Process XML content with deduplication
    processed_content, duplicates_removed = process_xml_with_deduplication(cleaned_content)
    final_content_parts.append(processed_content)
    
    final_content = ''.join(final_content_parts)
    
    # Write output
    output_filename = get_output_filename(args.file, args.replace)
    try:
        with open(output_filename, 'w', encoding='utf-8') as f:
            f.write(final_content)
        
        # Set file permissions (Unix/Linux)
        if hasattr(os, 'chmod'):
            os.chmod(output_filename, FILE_PERMISSIONS)
            
    except Exception as e:
        print(f"Error: Could not write output file: {e}", file=sys.stderr)
        sys.exit(1)
    
    # Handle file replacement
    if args.replace:
        try:
            os.rename(output_filename, args.file)
            print(f"Original file replaced: {args.file}", end="")
        except Exception as e:
            try:
                os.remove(output_filename)
            except:
                pass
            print(f"Error: Could not replace original file: {e}", file=sys.stderr)
            sys.exit(1)
    else:
        print(f"Organized project saved to: {output_filename}", end="")
    
    if duplicates_removed > 0:
        print(f" (removed {duplicates_removed} duplicates)", end="")
    
    mode_text = " (with logical organization)" if args.organize else " (preserving original structure)"
    print(mode_text)

def main():
    """Main entry point with error handling"""
    try:
        args = parse_args()
        process_file(args)
    except KeyboardInterrupt:
        print("", file=sys.stderr)  # Clean newline on Ctrl+C
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()